cloud/pkg/controllermanager/edgeapplication/edgeapplicationcontroller.go

```diff
@@ -122,20 +125,52 @@ func (c *Controller) syncEdgeApplication(ctx context.Context, edgeApp *appsv1alp
 			continue
 		}
 
-		// apply overriders
-		//
-		// TODO: consider the situation that not all the overrides have been applied successfully
-		// If one succeeded and another failed, the status of edgeApp will only contain the successful
-		// one, and have no status about the failed one.
 		for _, info := range overriderInfos {
 			tmplCopy := tmpl.DeepCopy()
-			klog.V(4).Infof("override obj %s/%s of gvk %s, for nodegroup %s", tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), info.TargetNodeGroup)
-			if err := c.Overrider.ApplyOverrides(tmplCopy, info); err != nil {
-				klog.Errorf("failed to apply override of nodegroup %s to obj %s/%s of gvk %s, %v",
-					info.TargetNodeGroup, tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), err)
-				errs = append(errs, err)
-				continue
+
+			// Handle node group overrides
+			if info.TargetNodeGroup != "" {
+				klog.V(4).Infof("override obj %s/%s of gvk %s, for nodegroup %s", tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), info.TargetNodeGroup)
+				if err := c.Overrider.ApplyOverrides(tmplCopy, info); err != nil {
+					klog.Errorf("failed to apply override of nodegroup %s to obj %s/%s of gvk %s, %v",
+						info.TargetNodeGroup, tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), err)
+					errs = append(errs, err)
+					continue
+				}
 			}
+
+			// Handle node label overrides
+			if info.TargetNodeLabelSelector.MatchLabels != nil {
+				nodes, err := utils.GetNodesByLabels(ctx, c.Client, info.TargetNodeLabelSelector.MatchLabels)
+				if err != nil {
+					klog.Errorf("failed to get nodes by labels for obj %s/%s of gvk %s, %v",
+						tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), err)
+					errs = append(errs, err)
+					continue
+				}
+
+				if len(nodes) > 0 {
+					klog.V(4).Infof("override obj %s/%s of gvk %s, for node labels", tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind())
+					if err := c.Overrider.ApplyOverrides(tmplCopy, info); err != nil {
+						klog.Errorf("failed to apply override for node labels to obj %s/%s of gvk %s, %v",
+							tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), err)
+						errs = append(errs, err)
+						continue
+					}
+
+					if err := utils.ApplyNodeAffinity(tmplCopy, info.TargetNodeLabelSelector); err != nil {
+						klog.Errorf("failed to apply node affinity to obj %s/%s of gvk %s, %v",
+							tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind(), err)
+						errs = append(errs, err)
+						continue
+					}
+				} else {
+					klog.Warningf("No nodes found matching labels for obj %s/%s of gvk %s",
+						tmplCopy.GetNamespace(), tmplCopy.GetName(), tmplCopy.GroupVersionKind())
+					continue
```

Comment:
It would be best to write the message to the EdgeApplication Status. And the historical error handling doesn't look  good, the Reconcile() function returns an error only make it constant retried.